lib/events/api.go

```diff
@@ -799,15 +799,29 @@ type UploadMetadataGetter interface {
 	GetUploadMetadata(sid session.ID) UploadMetadata
 }
 
-// StreamWriter implements io.Writer to be plugged into the multi-writer
-// associated with every session. It forwards session stream to the audit log
-type StreamWriter interface {
+// SessionEventSetter will set necessary event fields for session-related
+// events and must be called before the event is recorded, regardless
+// of whether the event will be recorded, emitted, or both.
+type SessionEventSetter interface {
+	SetupSessionEvent(event apievents.AuditEvent) (apievents.AuditEvent, error)
+}
```

Comment:
Addressed both suggestions. I had thought before about somehow enforcing that event setup is done before recording but this seems like a clear way to do it. I opted to make `PreparedSessionEvent` a new interface so it can't accidentally be misused by simply casting a `AuditEvent`. `SetupSessionEvent` being separate is a necessary evil unfortunately.

Previously `SessionWriter` would set event fields in the caller's goroutine, and `AsyncEmitter` would check the same event fields in a new goroutine when events were recorded or emitted respectively. It was very easy to trigger a data race by emitting and recording an event in that order. Making event setup separate means the same event can be recorded and emitted in either order with no problems.